README
NAME
    PlugAuth::Client - PlugAuth Client

VERSION
    version 0.20

SYNOPSIS
    In a perl program :

     my $r = PlugAuth::Client->new;

     # Check auth server status and version
     my $check = $r->status;
     my $version = $r->version;

     # Authenticate user "alice", pw "sesame"
     $r->login("alice", "sesame");
     if ($r->auth) {
        print "authentication succeeded\n";
     } else {
        print "authentication failed\n";
     }

     # Authorize "alice" to "POST" to "/board"
     if ($r->authz("alice","POST","board")) {
         print "authorization succeeded\n";
     } else {
         print "authorization failed\n";
     }

DESCRIPTION
    This module provides a perl front-end to the PlugAuth API. For a
    stripped down interface to just the authentication and authorization API
    (that is not including the user/group/authorization management
    functions), see PlugAuth::Client::Tiny.

METHODS
  $client->auth
    Returns true if the PlugAuth server can authenticate the user. Username
    and passwords can be specified with the login method or via the
    application's configuration file, see Clustericious::Client for details.

  $client->authz($user $action, $resource)
    Returns true if the given user ($user) is authorized to perform the
    given action ($action) on the given resource ($resource).

  $client->user
    Returns a list reference containing all usernames.

  $client->create_user( \%args )
    Create a user with the given username and password.

    *   user

        The new user's username

        REQUIRED

    *   password

        The new user's password

        REQUIRED

    *   groups

        List of groups as a comma separated string. Using this option
        requires that the server is running PlugAuth 0.21 or better.

  $client->delete_user( $username )
    Delete the user with the given username.

  $client->groups($user)
    Returns a list reference containing the groups that the given user
    ($user) belongs to.

  $client->change_password($user, $password)
    Change the password of the given user ($user) to a new password
    ($password).

  $client->group
    Returns a list reference containing all group names.

  $client->users($group)
    Returns the list of users belonging to the given group ($group).

  $client->create_group( \%args )
    Create a group.

    *   group

        The name of the new group

    *   users

        Comma separated list (as a string) of the users that should
        initially belong to this group.

  $client->update_group( $group, '--users' => $users )
    Update the given group ($group) replacing the existing list with the new
    list ($users), which is a comma separated list as a string.

  $client->delete_group( $group )
    Delete the given group ($group).

  $client->group_add_user( $group, $user )
    Adds the given user ($user) to the given group ($group)

  $client->group_delete_user( $group, $user )
    Delete the given user ($user) from the given group ($group)

  $client->grant( $user, $action, $resource )
    Grants the given user ($user) the authorization to perform the given
    action ($action) on the given resource ($resource).

  $client->revoke( $user, $action, $resource )
    Revokes permission for the give user ($user) to perform the given action
    ($action) on the given resource ($resource).

  $client->granted
    Returns a list of granted permissions

  $client->actions
    Returns a list reference containing the actions that the PlugAuth server
    knows about.

  $client->host_tag($ip_address, $tag)
    Returns true if the host specified by the given IP address ($ip_address)
    has the given host tag ($tag).

  $client->resources( $user, $action, $resource_regex )
    Returns a list reference containing the resources that match the regex
    provided ($resource_regex) that the given user ($user) can perform the
    given action ($action). To see all the resources that the user can
    perform the given action against, pass in '.*' as the regex.

  $client->action_resources( $user )
    Returns a hash reference of all actions and resources that the given
    user ($user) can perform. The keys in the returned hash are the actions
    and the values are list references containing the resources where those
    actions can be performed by the user.

  $client->audit( $year, $month, $day )
    Interface to the Audit plugin, if it is available.

SEE ALSO
    Clustericious::Client, PlugAuth, plugauthclient, plugauthpasswd,
    PlugAuth::Client::Tiny

AUTHOR
    Graham Ollis <gollis@sesda3.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.20      2013-07-19 16:23:44 America/New_York
  - documentation fixes

0.19_01   2013-07-19 10:11:21 America/New_York
  - support creating users with

0.19      2013-06-30 19:19:23 America/New_York
  - use YAML::XS instead of YAML, since Clustericious requires YAML::XS (but not YAML)

0.18      2013-05-09 23:11:15 America/New_York
  - Documentation fixes.

0.17      2013-05-09 21:10:50 America/New_York
  - Documentation fixes.

0.16      2013-04-12 11:41:18 America/New_York
  - added route for audit log (requires server using the Audit plugin)

0.15      2012-12-03 10:51:56 America/New_York
  - provide granted method (GET /grat)

0.14      2012-11-29 14:36:29 America/New_York
  - use modifies_payload

0.13      2012-11-29 11:56:26 America/New_York
  - fixed bug with grant on resource /
  - renamed to PlugAuth/plugauth
  - added revoke route (DELETE /grant/...)

0.11      2012-10-26 09:53:34 America/New_York
  - remove dupes from action_resource

0.10      2012-10-18 16:21:49 America/New_York
  - add action_resources method

0.09      2012-10-17 15:56:51 America/New_York
  - add change_password method


eg/test.pl
#!/usr/bin/env perl

use PlugAuth::Client;
use Data::Dumper;

my $r = PlugAuth::Client->new();

print Dumper($r->status);
print Dumper($r->version);

$r->login(username => "charliebrown", password => "snoopy");
if ($r->auth) {
    print "ok\n";
} else {
    print "failed\n";
}

$r->login(username => "charliebrown", password => "snoopy2");
if ($r->auth) {
    print "ok\n";
} else {
    print "failed\n";
}

if ($r->authz("charliebrown","kick","football")) {
    print "ok\n";
} else {
    print "failed\n";
}

if ($r->authz("charliebrown","kick","flootball")) {
    print "ok\n";
} else {
    print "failed\n";
}


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "PlugAuth Client",
  "dist_author" => [
    "Graham Ollis <gollis\@sesda3.com>"
  ],
  "dist_name" => "PlugAuth-Client",
  "dist_version" => "0.16",
  "license" => "restrictive",
  "module_name" => "PlugAuth::Client",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Clustericious::Client" => "0.73",
    "Clustericious::Client::Command" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "Log::Log4perl" => 0,
    "Log::Log4perl::CommandLine" => 0,
    "Term::Prompt" => 0,
    "YAML" => 0,
    "perl" => "v5.10.0"
  },
  "script_files" => [
    "bin/plugauthclient",
    "bin/plugauthpasswd"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "PlugAuth Client",
  "AUTHOR" => "Graham Ollis <gollis\@sesda3.com>",
  "BUILD_REQUIRES" => {
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "PlugAuth-Client",
  "EXE_FILES" => [
    "bin/plugauthclient",
    "bin/plugauthpasswd"
  ],
  "LICENSE" => "perl",
  "NAME" => "PlugAuth::Client",
  "PREREQ_PM" => {
    "Clustericious::Client" => "0.73",
    "Clustericious::Client::Command" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "Log::Log4perl" => 0,
    "Log::Log4perl::CommandLine" => 0,
    "Term::Prompt" => 0,
    "YAML::XS" => 0
  },
  "VERSION" => "0.20",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/PlugAuth/Client.pm
package PlugAuth::Client;

use strict;
use warnings;
use v5.10;
use Log::Log4perl qw(:easy);
use Clustericious::Client;

# ABSTRACT: PlugAuth Client
our $VERSION = '0.20'; # VERSION


route welcome      => 'GET', '/';


route auth         => 'GET', '/auth';


route_doc authz    => "user action resource";
sub authz
{
  my($self, $user, $action, $resource) = @_;

  my $url = Mojo::URL->new( $self->server_url );

  $resource = "/$resource" unless $resource =~ m{^/};
  
  $url->path("/authz/user/$user/$action$resource");

  $self->_doit('GET', $url);
}


route user         => 'GET', '/user';


route create_user => 'POST', '/user', \("--user username --password password");
route_args create_user => [
  { name => 'user',     type => '=s', required => 1, modifies_payload => 'hash' },
  { name => 'password', type => '=s', required => 1, modifies_payload => 'hash' },
  { name => 'groups',   type => '=s', required => 0, modifies_payload => 'hash' },
];


route delete_user  => 'DELETE',  '/user', \("user");
route_args delete_user => [
  { name => 'user', type => '=s', required => 1, modifies_url => 'append' },
];


route groups       => 'GET', '/groups', \("user");


route_doc change_password => 'username password';
sub change_password
{
  my($self, $user, $password) = @_;
  my $url = Mojo::URL->new( $self->server_url );
  $url->path("/user/$user");
  $self->_doit('POST', $url, { password => $password });
}


route group        => 'GET', '/group';


route users        => 'GET', '/users', \("group");


route create_group => 'POST', '/group', \("--group group --users user1,user2,...");
route_args create_group => [
  { name => 'group', type => '=s', required => 1, modifies_payload => 'hash'  },
  { name => 'users', type => '=s', required => 1, modifies_payload => 'hash'  },
];


route_doc 'update_group' => 'group --users user1,user2,...';
route_args update_group => [
  { name => 'group', type => '=s', required => 1, modifies_url => 'append', 'positional' => 'one' },
  { name => 'users', type => '=s', required => 1 },
];
sub update_group
{
  my $self = shift;
  my $group = shift;
  my $args = ref($_[0]) eq 'HASH' ? $_[0] : {@_}; 

  LOGDIE "group needed for update"
    unless $group;

  my $url = Mojo::URL->new( $self->server_url );
  $url->path("/group/$group");

  TRACE("updating $group ", $url->to_string);

  $self->_doit('POST', $url, { users => $args->{users} // $args->{'--users'} });
}


route delete_group => 'DELETE', '/group', \("group");


route 'group_add_user' => 'POST' => '/group';
route_args 'group_add_user' => [
  { name => 'group', type => '=s', modifies_url => 'append', 'positional' => 'one' },
  { name => 'user',  type => '=s', modifies_url => 'append', 'positional' => 'one' },
];


route 'group_delete_user' => 'DELETE' => '/group';
route_args 'group_delete_user' => [
  { name => 'group', type => '=s', modifies_url => 'append', 'positional' => 'one' },
  { name => 'user',  type => '=s', modifies_url => 'append', 'positional' => 'one' },
];


route 'grant' => 'POST' => '/grant';
route_args 'grant' => [
  { name => 'user',     type => '=s', modifies_url => 'append', positional => 'one' },
  { name => 'action',   type => '=s', modifies_url => 'append', positional => 'one' },
  { name => 'resource', type => '=s', modifies_url => 'append', positional => 'one' },
];


route 'revoke' => 'DELETE' => '/grant';
route_args 'revoke' => [
  { name => 'user',     type => '=s', modifies_url => 'append', positional => 'one' },
  { name => 'action',   type => '=s', modifies_url => 'append', positional => 'one' },
  { name => 'resource', type => '=s', modifies_url => 'append', positional => 'one' },
];


route granted      => 'GET', '/grant';


route actions      => 'GET', '/actions';


route host_tag     => 'GET', '/host', \("host tag");


route resources    => 'GET', '/authz/resources', \("user action resource_regex");


sub _remove_prefixes
{
  my @in = sort @_;
  my @out;
  while(my $item = shift @in)
  {
    @in = grep { substr($_, 0, length $item) ne $item } @in;
    push @out, $item;
  }
  @out;
}

route_doc action_resources => "user";
sub action_resources
{
  my($self, $user) = @_;
  my %table;
  foreach my $action (@{ $self->actions })
  {
    my $resources = $self->resources($user, $action, '/');
    $table{$action} = [_remove_prefixes(@$resources)] if @$resources > 0;
  }
  \%table;
}


route_doc action_resource => 'audit';
sub audit
{
  my($self, $year, $month, $day) = @_;
  my $uri;
  if(defined $day)
  {
    $uri = join('/', '', 'audit', $year, sprintf("%02d", $month), sprintf("%02d", $day));
  }
  else
  {
    # TODO: Clustericious::Client doesn't handle 302 correctly
    $uri = join('/', '', 'audit', 'today');
  }
  $self->_doit(GET => $uri);
}

1;



=pod

=head1 NAME

PlugAuth::Client - PlugAuth Client

=head1 VERSION

version 0.20

=head1 SYNOPSIS

In a perl program :

 my $r = PlugAuth::Client->new;

 # Check auth server status and version
 my $check = $r->status;
 my $version = $r->version;

 # Authenticate user "alice", pw "sesame"
 $r->login("alice", "sesame");
 if ($r->auth) {
    print "authentication succeeded\n";
 } else {
    print "authentication failed\n";
 }

 # Authorize "alice" to "POST" to "/board"
 if ($r->authz("alice","POST","board")) {
     print "authorization succeeded\n";
 } else {
     print "authorization failed\n";
 }

=head1 DESCRIPTION

This module provides a perl front-end to the PlugAuth API.  For a stripped
down interface to just the authentication and authorization API (that is
not including the user/group/authorization management functions), see
L<PlugAuth::Client::Tiny>.

=head1 METHODS

=head2 $client-E<gt>auth

Returns true if the PlugAuth server can authenticate the user.  
Username and passwords can be specified with the login method or
via the application's configuration file, see L<Clustericious::Client>
for details.

=head2 $client-E<gt>authz($user $action, $resource)

Returns true if the given user ($user) is authorized to perform the
given action ($action) on the given resource ($resource).

=head2 $client-E<gt>user

Returns a list reference containing all usernames.

=head2 $client-E<gt>create_user( \%args )

Create a user with the given username and password.

=over 4

=item * user

The new user's username

REQUIRED

=item * password

The new user's password

REQUIRED

=item * groups

List of groups as a comma separated string.  Using this option requires that
the server is running PlugAuth 0.21 or better.

=back

=head2 $client-E<gt>delete_user( $username )

Delete the user with the given username.

=head2 $client-E<gt>groups($user)

Returns a list reference containing the groups that the given user ($user)
belongs to.

=head2 $client-E<gt>change_password($user, $password)

Change the password of the given user ($user) to a new password ($password).

=head2 $client-E<gt>group

Returns a list reference containing all group names.

=head2 $client-E<gt>users($group)

Returns the list of users belonging to the given group ($group).

=head2 $client-E<gt>create_group( \%args )

Create a group.

=over 4

=item * group

The name of the new group

=item * users

Comma separated list (as a string) of the users that
should initially belong to this group.

=back

=head2 $client-E<gt>update_group( $group, '--users' => $users )

Update the given group ($group) replacing the existing list with
the new list ($users), which is a comma separated list as a string.

=head2 $client-E<gt>delete_group( $group )

Delete the given group ($group).

=head2 $client-E<gt>group_add_user( $group, $user )

Adds the given user ($user) to the given group ($group)

=head2 $client-E<gt>group_delete_user( $group, $user )

Delete the given user ($user) from the given group ($group)

=head2 $client-E<gt>grant( $user, $action, $resource )

Grants the given user ($user) the authorization to perform the given
action ($action) on the given resource ($resource).

=head2 $client-E<gt>revoke( $user, $action, $resource )

Revokes permission for the give user ($user) to perform the given action ($action)
on the given resource ($resource).

=head2 $client-E<gt>granted

Returns a list of granted permissions

=head2 $client-E<gt>actions

Returns a list reference containing the actions that the PlugAuth server
knows about.

=head2 $client-E<gt>host_tag($ip_address, $tag)

Returns true if the host specified by the given IP address ($ip_address)
has the given host tag ($tag).

=head2 $client-E<gt>resources( $user, $action, $resource_regex )

Returns a list reference containing the resources that match the regex
provided ($resource_regex) that the given user ($user) can perform the
given action ($action).  To see all the resources that the user can
perform the given action against, pass in '.*' as the regex.

=head2 $client-E<gt>action_resources( $user )

Returns a hash reference of all actions and resources that the given
user ($user) can perform.  The keys in the returned hash are the 
actions and the values are list references containing the resources
where those actions can be performed by the user.

=head2 $client-E<gt>audit( $year, $month, $day )

Interface to the L<Audit|PlugAuth::Plugin::Audit> plugin, if it is available.

=head1 SEE ALSO

L<Clustericious::Client>, 
L<PlugAuth>, 
L<plugauthclient>, 
L<plugauthpasswd>,
L<PlugAuth::Client::Tiny>

=head1 AUTHOR

Graham Ollis <gollis@sesda3.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__

